src/Pipeline.cpp

```diff
@@ -673,42 +689,50 @@ struct JITFuncCallContext {
 
 }  // namespace
 
+struct Pipeline::JITCallArgs {
+    size_t size{0};
+    const void *fixed_store[64];
+    const void **store;
+
+    JITCallArgs(size_t size) : size(size) {
+        if (size > (sizeof(fixed_store) / sizeof(fixed_store[0]))) {
+            // TODO(zalman): Call new[]?
+            store = (const void **)malloc(sizeof(void *) * size);
+        } else {
+            store = fixed_store;
+        }
+    }
+
+  ~JITCallArgs() {
+      if (store != fixed_store) {
+          free(store);
+      }
+    }
+};
+
 // Make a vector of void *'s to pass to the jit call using the
 // currently bound value for all of the params and image
 // params.
-vector<const void *> Pipeline::prepare_jit_call_arguments(Realization dst, const Target &target,
-                                                          const ParamMap &param_map, void *user_context,
-                                                          bool is_bounds_inference) {
+void Pipeline::prepare_jit_call_arguments(Realization dst, const Target &target,
+                                          const ParamMap &param_map, void *user_context,
+                                          bool is_bounds_inference, JITCallArgs &args_result) {
     user_assert(defined()) << "Can't realize an undefined Pipeline\n";
 
     compile_jit(target);
 
     JITModule &compiled_module = contents->jit_module;
     internal_assert(compiled_module.argv_function());
 
-    struct OutputBufferType {
-        Function func;
-        Type type;
-        int dims;
-    };
-    vector<OutputBufferType> output_buffer_types;
-    for (Function f : contents->outputs) {
-        for (Type t : f.output_types()) {
-            OutputBufferType obt = {f, t, f.dimensions()};
-            output_buffer_types.push_back(obt);
-        }
-    }
-
-    user_assert(output_buffer_types.size() == dst.size())
+    user_assert(contents->output_buffer_types.size() == dst.size())
         << "Realization contains wrong number of Images (" << dst.size()
-        << ") for realizing pipeline with " << output_buffer_types.size()
+        << ") for realizing pipeline with " << contents->output_buffer_types.size()
         << " outputs\n";
 
     // Check the type and dimensionality of the buffer
     for (size_t i = 0; i < dst.size(); i++) {
-        Function func = output_buffer_types[i].func;
-        int  dims = output_buffer_types[i].dims;
-        Type type = output_buffer_types[i].type;
+        Function &func = contents->output_buffer_types[i].func;
+        int  dims = contents->output_buffer_types[i].dims;
+        Type &type = contents->output_buffer_types[i].type;
         user_assert(dst[i].dimensions() == dims)
```

Comment:
Dimensionality and type is now checked inside the kernel - I think it wasn't when this code was written. This loop can probably be cut entirely.... do we even need output_buffer_types?

Reply:
Is it checked if no_asserts is on? Should we keep this code for that case? (I can go either way in that it seems prudent, on the other hand, no_asserts definitely gives us license to not do the check and should save as many cycles as possible.) The cost of making output_buffer_types now is minimal, but checking it twice is certainly not useful.

Reply:
Good point, it's not. But yeah if someone is using no_asserts they presumably don't want this sort of assert. I'd cut it regardless.